[GTK][WPE] Simplify damage tracking implementation

carlosgcampos opened the pull request:
#### d4f357221ebcba32b6c44fe274654c3c387ec810
<pre>
[GTK][WPE] Simplify damage tracking implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=283752">https://bugs.webkit.org/show_bug.cgi?id=283752</a>

Reviewed by Adrian Perez de Castro.

Now that we collect damage before the actual painting, the code can be
simplified. We can also remove the amount of ifdefs. This patch adds
ENABLE(DAMAGE_TRACKING) to replace ENABLE(WPE_PLATFORM) || PLATFORM(GTK).

* Source/WTF/wtf/PlatformEnableGlib.h:
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::invalidate):
(WebCore::Damage::Damage):
* Source/WebCore/platform/graphics/nicosia/NicosiaCompositionLayer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::setDamage):
(WebCore::TextureMapperLayer::collectDamage):
(WebCore::TextureMapperLayer::collectDamageRecursive):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::transformRectForDamage):
(WebCore::TextureMapperLayer::addChild):
(WebCore::TextureMapperLayer::removeFromParent):
(WebCore::TextureMapperLayer::removeAllChildren):
(WebCore::TextureMapperLayer::applyAnimationsRecursively):
(WebCore::TextureMapperLayer::TextureMapperLayer): Deleted.
(WebCore::TextureMapperLayer::acceptDamageVisitor): Deleted.
(WebCore::TextureMapperLayer::dismissDamageVisitor): Deleted.
(WebCore::TextureMapperLayer::recordDamage): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::clearDamage): Deleted.
(WebCore::TextureMapperLayer::invalidateDamage): Deleted.
(WebCore::TextureMapperLayer::addDamage): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::setContentsOpaque):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplay):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplayInRect):
(WebCore::CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::didRenderFrame):
(WebKit::AcceleratedSurface::addDamage):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::CoordinatedGraphicsScene):
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):
(WebKit::texmapLayer):
(WebKit::CoordinatedGraphicsScene::updateSceneState):
(WebKit::CoordinatedGraphicsScene::ensureRootLayer):
(WebKit::CoordinatedGraphicsScene::purgeGLResources):
(WebKit::CoordinatedGraphicsScene::recordDamage): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.h:
(WebKit::CoordinatedGraphicsScene::setDamagePropagation):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::LayerTreeHost):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::setDamagePropagation):
(WebKit::ThreadedCompositor::addSurfaceDamage):
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::addDamage):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.cpp:
(WebKit::AcceleratedSurfaceLibWPE::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.h:

Canonical link: <a href="https://commits.webkit.org/287146@main">https://commits.webkit.org/287146@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54d9eabc5400cd606517d93a3f99a20d1da40fdb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78520 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57565 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31902 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83181 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29785 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80653 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5846 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61501 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/19416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81587 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51513 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69612 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41813 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48859 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25316 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28122 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71666 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69958 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25689 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84547 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77757 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5885 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4024 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/69727 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6046 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/67490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/68981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13001 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11423 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100066 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12125 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5832 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/10615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21855 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5820 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9254 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7607 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Scony commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -81,29 +67,28 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
-            if (!damageSinceLastSurfaceUse.isInvalid())
-                actualClipRect = static_cast<FloatRoundedRect>(damageSinceLastSurfaceUse.bounds());
+
+        const auto& damageSinceLastSurfaceUse = m_client->addSurfaceDamage(frameDamage);
```
this should be happening regardless of `sceneHasRunningAnimations` as otherwise it may not be called some frames and we may end up with wrong per-RenderTarget-damage

carlosgcampos replied:
Fixed

Scony reviewed (approved):
LGTM

aperezdc commented on Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
```diff
@@ -673,11 +673,13 @@ void AcceleratedSurfaceDMABuf::didRenderFrame(WebCore::Region&& damage)
 
 const WebCore::Damage& AcceleratedSurfaceDMABuf::addDamage(const WebCore::Damage& damage)
 {
+    m_frameDamage = damage;
```
Wouldn't this be `m_frameDamage.add(damage)`? Currently if there is more than one call to `AcceleratedSurfaceDMABuf::addDamage()` only the damage passed in the last call is recorded.

```suggestion
    m_frameDamage.add(damage);
```

carlosgcampos replied:
I don't think so, because this is the damage for this particular frame, what we passed to didRenderFrame before. The accumulated damage is handled by the render target already.

aperezdc commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -355,12 +348,76 @@ void TextureMapperLayer::paint(TextureMapper& textureMapper)
+
+void TextureMapperLayer::collectDamageRecursive(TextureMapperPaintOptions& options, Damage& damage)
+{
+    if (!isVisible())
+        return;
```
I think it should be possible to return early here if the `damage` passed as argument is already invalid, because adding to it is a no-op and it would still be invalid&mdash;and we would avoid a subtree traversal. We have to be careful to reset `m_damage = Damage()` if we return early because that is currently done inside `::collectDamageSelf()`.

Maybe something like this:

```suggestion
    if (!isVisible())
        return;
        
    if (damage.isInvalid()) {
        m_damage = Damage();
        return;
    }
```

carlosgcampos replied:
Don't we need to clear damage in the descendants too?

carlosgcampos replied:
And I think it's never invalid

aperezdc replied:
Right, I double checked and that's the case because `::collectDamage()` happens to be called always with a fresh `Damage` instance in `CoordinatedGraphicsScene::paintToCurrentGLContext()`. But then there's a check that we can remove from there, I'll add another comment in the proper location.

aperezdc reviewed (commented):
Very nice simplification! I have a couple of small comments.

aperezdc commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -81,29 +67,31 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
+            currentRootLayer->collectDamage(*m_textureMapper, frameDamage);
+            WTFEndSignpost(this, CollectDamage);
+
+            ASSERT(!frameDamage.isInvalid());
```
Aye, we are calling `TextureMapperLayer::collectDamage()` always with a fresh `Damage`, and that itself won't invalidate the instance. Having the assertion here documents this expectation and then the check inside the `if` in the next line is fine to remove.

aperezdc reviewed (approved):
r=me 🖌️

carlosgcampos committed:
[GTK][WPE] Simplify damage tracking implementation
https://bugs.webkit.org/show_bug.cgi?id=283752

Reviewed by Adrian Perez de Castro.

Now that we collect damage before the actual painting, the code can be
simplified. We can also remove the amount of ifdefs. This patch adds
ENABLE(DAMAGE_TRACKING) to replace ENABLE(WPE_PLATFORM) || PLATFORM(GTK).

* Source/WTF/wtf/PlatformEnableGlib.h:
* Source/WebCore/platform/graphics/Damage.h:
(WebCore::Damage::invalidate):
(WebCore::Damage::Damage):
* Source/WebCore/platform/graphics/nicosia/NicosiaCompositionLayer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::setDamage):
(WebCore::TextureMapperLayer::collectDamage):
(WebCore::TextureMapperLayer::collectDamageRecursive):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::transformRectForDamage):
(WebCore::TextureMapperLayer::addChild):
(WebCore::TextureMapperLayer::removeFromParent):
(WebCore::TextureMapperLayer::removeAllChildren):
(WebCore::TextureMapperLayer::applyAnimationsRecursively):
(WebCore::TextureMapperLayer::TextureMapperLayer): Deleted.
(WebCore::TextureMapperLayer::acceptDamageVisitor): Deleted.
(WebCore::TextureMapperLayer::dismissDamageVisitor): Deleted.
(WebCore::TextureMapperLayer::recordDamage): Deleted.
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::clearDamage): Deleted.
(WebCore::TextureMapperLayer::invalidateDamage): Deleted.
(WebCore::TextureMapperLayer::addDamage): Deleted.
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::setContentsOpaque):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplay):
(WebCore::CoordinatedGraphicsLayer::setNeedsDisplayInRect):
(WebCore::CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebKit/WebProcess/WebPage/AcceleratedSurface.h:
(WebKit::AcceleratedSurface::didRenderFrame):
(WebKit::AcceleratedSurface::addDamage):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::CoordinatedGraphicsScene):
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):
(WebKit::texmapLayer):
(WebKit::CoordinatedGraphicsScene::updateSceneState):
(WebKit::CoordinatedGraphicsScene::ensureRootLayer):
(WebKit::CoordinatedGraphicsScene::purgeGLResources):
(WebKit::CoordinatedGraphicsScene::recordDamage): Deleted.
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.h:
(WebKit::CoordinatedGraphicsScene::setDamagePropagation):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::LayerTreeHost):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.cpp:
(WebKit::ThreadedCompositor::setDamagePropagation):
(WebKit::ThreadedCompositor::addSurfaceDamage):
(WebKit::ThreadedCompositor::renderLayerTree):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/ThreadedCompositor.h:
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::didRenderFrame):
(WebKit::AcceleratedSurfaceDMABuf::addDamage):
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.h:
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.cpp:
(WebKit::AcceleratedSurfaceLibWPE::didRenderFrame):
* Source/WebKit/WebProcess/WebPage/libwpe/AcceleratedSurfaceLibWPE.h:

Canonical link: https://commits.webkit.org/287146@main
